Add: 카카오 회원가입 및 로그인 기능 구현

palza4dev opened the pull request:
- [x] 기능 추가
- [x] 리뷰 반영
- [x] 리팩토링
- [ ] 버그 수정
- [x] 컨벤션 수정

<br />

## :: 구현 목표 (해당 브랜치에서 구현하고자 하는 하나의 목표를 설정합니다.)
- 카카오 소셜 로그인을 이용한 회원가입 및 로그인 기능 구현

<br />

## :: 구현 사항 설명 (작업한 내용을 상세하게 기록합니다.)
1. 카카오 토큰을 받아서 카카오 API로 사용자 정보 요청
   - 카카오 API URL로 토큰을 보내어 해당 사용자의 정보를 받기

2. `get_or_create`로 기존 고객인지 판단
   - 기준은 kakao_id 로만 판단하고, 필수정보인 nickname과 image는 defaults 설정함
   - 신규회원이면 create하여 회원가입

3. 선택 동의에 따른 정보 수집
   - 사용자 정보 응답에서 email 혹은 age_range가 있는지 여부에 따라 DB에 추가
   
4.  유닛테스트 코드 작성
    - MagicMock과 patch를 활용해 unittest 진행
    - 성공 케이스: 1. 토큰을 생성후 전달 2. 다시 토큰을 받아서 DB에 유저를 확인 
    - 실패 케이스: 카카오에 요청시 토큰 확인이 안되는 경우

<br />

## :: 성장 포인트 (해당 기능을 구현하며 고민했던 사항이나 새로 알게된 부분, 어려웠던 점 등을 작성합니다.)
- 프론트-백엔드-카카오 통신 과정에서 디버깅을 하기 위해  `print(response)` 를 활용했고 어떤 부분이 문제였는지 파악 할 수 있었습니다.

- 카카오 API URL 주소는 https로 시작되는데 http와 어떤 차이인지 알아보고 싶습니다.

- 유닛테스트의 목적과 개념은 이해했지만, 구체적으로 어떻게 작동되는지는 아직 어렵습니다.

- 카카오 API를 따로 관리하기 위해 kakao.py 생성하여 KakaoAPI class 작성

<br />

## :: 기타 질문 및 특이 사항
-  선택 동의 정보 항목들은 if문으로 처리했는데 더 효율적인 방법이 있을 지 궁금합니다.

- 유닛테스트시 status_code 관련 attribute 에러가 계속 발생했는데,
 MockedResponse class 에 __init__으로 status_code 속성을 만들어주었습니다. 맞는 방법인지 스스로 의심이 갑니다..


## :: 테스트 결과 사진 

- 유닛테스트
<img width="930" alt="스크린샷 2021-10-23 오후 3 59 52" src="https://user-images.githubusercontent.com/72376931/138623524-aeabdad5-4713-4afc-8864-f379aa19430c.png">

- 카카오 소셜 로그인으로 회원가입
<img width="1489" alt="스크린샷 2021-10-23 오후 4 04 05" src="https://user-images.githubusercontent.com/72376931/138623519-03b1c6ba-72e5-423c-b0c5-6faa3fc1300c.png">

lordmyshepherd commented on users/views.py:
```diff
@@ -1,3 +1,45 @@
+    
+            KAKAO_URL     = 'https://kapi.kakao.com/v2/user/me'
+            headers       = {'Authorization' : f'Bearer {kakao_token}'}
+            response      = requests.post(KAKAO_URL, headers=headers).json()
```
`requests` 모듈을 사용하실 때에는 `timeout` 옵션을 반드시 설정해주셔야 합니다.
그 이유는 timeout option을 추가하지 않으면 서버의 응답을 계속해서 기다리고 있는 문제가 발생 합니다. timeout 옵션을 추가해 주세요.

```python
예를 들어
class KakaoLoginView(View):
    def post(self, request):
        try:
            kakao_access_token    = request.headers['Authorization']
            kakao_user_api_url    = '<https://kapi.kakao.com/v2/user/me>'
            headers     = {"Authorization": f"Bearer ${kakao_access_token}"}

            response = requests.get(kakao_user_api_url, headers = headers, timeout = ..)

```

palza4dev replied:
requests 문서를 찾아보니, 거의 모든 경우에 timeout 옵션을 추가하라는 내용이 있었습니다. 새로운 모듈, 라이브러리를 사용할 땐 항상 공식 문서를 먼저 잘 읽어보겠습니다. 튜플로 입력하면 서버 연결과 응답 대기 시간 timeout을 각각 설정할 수 있다는 것도 알았습니다. `timeout=3`으로 설정하겠습니다.

lordmyshepherd commented on users/views.py:
```diff
@@ -1,3 +1,45 @@
+    
+            KAKAO_URL     = 'https://kapi.kakao.com/v2/user/me'
+            headers       = {'Authorization' : f'Bearer {kakao_token}'}
+            response      = requests.post(KAKAO_URL, headers=headers).json()
```
아래와 같이 Kakao API 관련된 로직은 class로 그룹핑할 수 있습니다.  

그렇게 되면 kakao API 관련된 로직은 하나의 class에서 관리하면 된다는 장점이 있고, 만약에 나중에 kakao api 호출부의 수정일 있을 경우 `KakaoLoginView`를 수정할 필요가 없이 `KakaoAPI` 클래스만 수정하면 됩니다. 

결국 Kakao API를 호출하는 위치에서는 Interface만 알고 사용하면 되고 내부적으로 어떻게 구현되었는지 알 필요가 없습니다.

```python
# kakao.py

class KakaoAPI:
    def __init__(self, access_token):
        self.access_token = access_token
        self.user_url     = 'https://kapi.kakao.com/v2/user/me'
	self.map_url      = 

    def get_kakao_user(self):
        headers = {"Authorization": f"Bearer ${self.access_token}"}

        response = requests.get(self.user_url, headers=headers)
    
	if not response.status_code == 200:
	          return
	
	return response.json()

    def get_location():
        return

class KakaoLoginView(View):
    def post(self, request):
        try:
            kakao_access_token = request.headers['Authorization']
            kakao_user = KakaoAPI(kakao_access_token)

```

palza4dev replied:
kakao API class를 나누어 관리하면 사용자 정보뿐 아니라 다양한 카카오 API를 활용하기 좋을 것 같습니다. 언제든지 카카오 로그인, 지도, 네비게이션, 카카오 페이까지 다양한 기능이 추가될 수 있도록 확장성을 고려해야겠습니다. 위 내용 고려해서 수정해보겠습니다!

lordmyshepherd commented on users/views.py:
```diff
@@ -1,3 +1,45 @@
+
+            if response['kakao_account']['age_range_needs_agreement'] == False:
+                user.age_range = response['kakao_account']['age_range']
+                user.save()
```
line 31 ~ 37까지 코드가 필요한가요?

lordmyshepherd commented on users/tests.py:
```diff
@@ -1,3 +1,63 @@
+		client = Client()
+
+		class MockedResponse:
+			def __init__(self, status_code):
```
해당 파일의 indentation이 참으로 맞지 않는 것 같다는 생각이 듭니다..

lordmyshepherd commented on users/tests.py:
```diff
@@ -1,3 +1,62 @@
+						"age_range_needs_agreement" : False,
+						"age_range" : "20-30"
+					}
+                }
```
indentation 맞춰주세요~

palza4dev committed:
Add: 카카오 회원가입 및 로그인 기능 구현
- kakao.py에 kakao API class 작성하여 분리하여 관리
- views.py에 KakaoSignUpView class 작
- unittest 완료
- 컨벤션 수정 및 indentation 에러 해결
- 선택동의정보 if문을 없애고 get_or_create 내부로 이동

lordmyshepherd commented on users/views.py:
```diff
@@ -1,3 +1,42 @@
+                defaults = {
+                    'nickname'     : nickname,
+                    'profile_image': profile_image
+                }
```
```suggestion
                defaults = {
                    'nickname'     : nickname,
                    'profile_image': profile_image,
                    'email' : kakao_account['profile'].get('email', ''),
                    'age_range' : kakao_account['profile'].get('age_range', ''),
                }
```

lordmyshepherd commented on users/views.py:
```diff
@@ -1,3 +1,42 @@
+
+            if kakao_account['age_range_needs_agreement'] == False:
+                user.age_range = kakao_account['age_range']
+                user.save()
```
```suggestion
```

lordmyshepherd reviewed (changes requested):
LGTM :th

lordmyshepherd commented:
LGTM 👍
